Alerting: Fix saving of screenshots uploaded with a signed url

VDVsx opened the pull request:
**What this PR does / why we need it**:

The URL of screenshots uploaded to external image storages can be optionally signed,
resulting in a long string (800+ chars).

Add a migration to support URLs with a max length of 2048 chars in alert_image table.

**Which issue(s) this PR fixes**:

Fixes #53566

**Special notes for your reviewer**:

In theory URL max length should be 2048 chars, this field should not be directly settable by the users so the risk of blowing up in size is minimal, but if there are concerns we can add a length check before saving to DB.

VDVsx committed:
Alerting: Fix saving of screenshots uploaded with a signed url

The URL of screenshots uploaded to external image storages can be optionally signed,
resulting in a long string (800+ chars).

Add a migration to change 'alert_image.url' to MEDIUMTEXT in MySQL and TEXT in Postgres.
SQLite is already defined by default as unlimited length TEXT.

Fixes #53566

Signed-off-by: Valério Valério <valerio.valerio@unity3d.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=53933) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=53933) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=53933) it.</sub>

VDVsx commented on pkg/services/ngalert/store/image_test.go:
```diff
@@ -90,8 +90,11 @@ func TestIntegrationGetImages(t *testing.T) {
 	assert.Equal(t, []models.Image{image1}, images)
 
-	// create an image with a URL
-	image2 := models.Image{Path: "https://example.com/example.png"}
```
This test was not creating an image with a URL as the comment said, so changed it to use an URL and made it a long one.

JohnnyQQQQ reviewed (dismissed):
I would discourage the use of any `text` type for MySQL as they are in separated files on disk and can introduce performance bottlenecks.

The common thing to do in MySQL for URLs is to use `VARCHAR(2083)` as this is the common denominator for the maximal URL length in browsers.

But I let @grobinson-grafana do the call here.

VDVsx committed:
Use VARCHAR(2083) instead of MEDIUMTEXT for MySQL

Signed-off-by: Valério Valério <valerio.valerio@unity3d.com>

VDVsx commented:
Good point,thanks. Text types are use in several places already for MySQL, but I do agree that there's no need in this case, changed.

grobinson-grafana commented on pkg/services/sqlstore/migrations/ualert/tables.go:
```diff
@@ -396,6 +396,11 @@ func AddAlertImageMigrations(mg *migrator.Migrator) {
 	mg.AddMigration("create alert_image table", migrator.NewAddTableMigration(imageTable))
 	mg.AddMigration("add unique index on token to alert_image table", migrator.NewAddIndexMigration(imageTable, imageTable.Indices[0]))
+
+	mg.AddMigration("alter alert_image.url to mediumtext in mysql and text in postgres", migrator.NewRawSQLMigration("").
```
```suggestion
	mg.AddMigration("support longer URLs in alert_image table", migrator.NewRawSQLMigration("").
		Default("ALTER TABLE alert_image ALTER COLUMN url TYPE VARCHAR(2048);").
		Mysql("ALTER TABLE alert_image MODIFY url VARCHAR(2048) NOT NULL;"))
```

VDVsx replied:
Thanks, good catch the nullable would have been dropped for MySQL :)

VDVsx replied:
I didn't use the `Default(...)` as suggested since SQLite does not support `ALTER COLUMN`, we would need to introduce more migration steps and since the column in SQLite already support long text it seems unnecessary. Let me know what you think.

grobinson-grafana commented on pkg/services/ngalert/store/image_test.go:
```diff
@@ -90,8 +90,11 @@ func TestIntegrationGetImages(t *testing.T) {
-	// create an image with a URL
-	image2 := models.Image{Path: "https://example.com/example.png"}
+	// create an image with a long URL (e.g signed URL)
+	image2 := models.Image{URL: "https://example.com/example.png?Signature=3r4r4r34r34r34r34dfeergef3t343t34erg443" +
```
Is the signature here just a random string?

VDVsx replied:
Yes, c&p 🐱 typing, I can change to something that looks more like a signature if that's preferable for aesthetics

VDVsx replied:
Ok, made it a little longer, over 1000 chars, this size is already longer than common signed URLs from the 3 major cloud providers.

VDVsx committed:
Change migration description as per PR review

Signed-off-by: Valério Valério <valerio.valerio@unity3d.com>

VDVsx committed:
Make the long url in ngalert test a little longer

The test url is now over 1000 chars.
GCS seems to be the one having longer signed URLs, a bit over
800 chars, AWS S3 and Azure blob storage produce signed URLs with
a bit over 300 chars.

Signed-off-by: Valério Valério <valerio.valerio@unity3d.com>

grobinson-grafana commented on pkg/services/sqlstore/migrations/ualert/tables.go:
```diff
@@ -396,6 +396,11 @@ func AddAlertImageMigrations(mg *migrator.Migrator) {
+
+	mg.AddMigration("support longer URLs in alert_image table", migrator.NewRawSQLMigration("").
+		Postgres("ALTER TABLE alert_image ALTER COLUMN url TYPE VARCHAR(2048);").
+		Mysql("ALTER TABLE alert_image MODIFY url VARCHAR(2048);"))
```
```suggestion
		Mysql("ALTER TABLE alert_image MODIFY url VARCHAR(2048) NOT NULL;"))
```

grobinson-grafana reviewed (changes requested):
This migration needs `NOT NULL` to prevent the column being nullable.

VDVsx committed:
Update pkg/services/sqlstore/migrations/ualert/tables.go

Co-authored-by: George Robinson <george.robinson@grafana.com>

grobinson-grafana commented:
Looks good to me!

```
MariaDB [grafana]> EXPLAIN alert_image;
+------------+---------------+------+-----+---------+----------------+
| Field      | Type          | Null | Key | Default | Extra          |
+------------+---------------+------+-----+---------+----------------+
| id         | bigint(20)    | NO   | PRI | NULL    | auto_increment |
| token      | varchar(190)  | NO   | UNI | NULL    |                |
| path       | varchar(190)  | NO   |     | NULL    |                |
| url        | varchar(2048) | NO   |     | NULL    |                |
| created_at | datetime      | NO   |     | NULL    |                |
| expires_at | datetime      | NO   |     | NULL    |                |
+------------+---------------+------+-----+---------+----------------+
6 rows in set (0.000 sec)
```

```
postgres=# \d alert_image
                                        Table "public.alert_image"
   Column   |            Type             | Collation | Nullable |                 D
efault
------------+-----------------------------+-----------+----------+------------------
-----------------------
 id         | integer                     |           | not null | nextval('alert_im
age_id_seq'::regclass)
 token      | character varying(190)      |           | not null |
 path       | character varying(190)      |           | not null |
 url        | character varying(2048)     |           | not null |
 created_at | timestamp without time zone |           | not null |
 expires_at | timestamp without time zone |           | not null |
Indexes:
    "alert_image_pkey" PRIMARY KEY, btree (id)
    "UQE_alert_image_token" UNIQUE, btree (token)
```

```
sqlite> .schema alert_image
CREATE TABLE `alert_image` (
`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL
, `token` TEXT NOT NULL
, `path` TEXT NOT NULL
, `url` TEXT NOT NULL
, `created_at` DATETIME NOT NULL
, `expires_at` DATETIME NOT NULL
);
CREATE UNIQUE INDEX `UQE_alert_image_token` ON `alert_image` (`token`);
```
